File: JAVASCRIPT/wallet.js
const { formatDate } = require('./utils');

class Wallet {
  constructor() {
    this.balance = 0;
    this.actions = [];
  }

  addMoney(amount) {
    this.balance += amount;
    this.actions.push({ type: 'Add Money', amount, balance: this.balance, date: new Date() });
  }

  withdrawMoney(amount) {
    if (amount > this.balance) {
      console.log("Insufficient balance to withdraw this amount.");
      return;
    }

    this.balance -= amount;
    this.actions.push({ type: 'Withdraw Money', amount, balance: this.balance, date: new Date() });
  }

  showBalance() {
    console.log(`Current balance: ${this.balance} MGA`);
  }

  displayHistory() {
    console.log("Action history:");
    this.actions.forEach((action, index) => {
      const formattedDate = formatDate(action.date);
      console.log(`${index + 1}. ${action.type}: ${action.amount} MGA (Balance: ${action.balance} MGA) - Date: ${formattedDate}`);
    });
  }
}

module.exports = Wallet;

File: PYTHON/wallet.py
from datetime import datetime
from utils import format_date

class Wallet:
    def __init__(self):
        self.balance = 0
        self.actions = []

    def add_money(self, amount):
        self.balance += amount
        self.actions.append({'type': 'Add Money', 'amount': amount, 'balance': self.balance, 'date': datetime.now()})

    def withdraw_money(self, amount):
        if amount > self.balance:
            print("Insufficient balance to withdraw this amount.")
            return

        self.balance -= amount
        self.actions.append({'type': 'Withdraw Money', 'amount': amount, 'balance': self.balance, 'date': datetime.now()})

    def show_balance(self):
        print(f"Current balance: {self.balance} MGA")

    def display_history(self):
        print("Action history:")
        for index, action in enumerate(self.actions, start=1):
            formatted_date = format_date(action['date'])
            print(f"{index}. {action['type']}: {action['amount']} MGA (Balance: {action['balance']} MGA) - Date: {formatted_date}")